lib/sycamore/sycamore/llms/openai.py

```diff
@@ -251,18 +254,45 @@ def __reduce__(self):
         def deserializer(kwargs):
             return OpenAI(**kwargs)
 
-        kwargs = {"client_wrapper": self.client_wrapper, "model_name": self.model_name}
+        kwargs = {"client_wrapper": self.client_wrapper, "model_name": self._model_name}
 
-        return (deserializer, (kwargs,))
+        return deserializer, (kwargs,)
 
     def is_chat_mode(self):
         return self.model.is_chat
 
+    @staticmethod
+    def _get_cache_key(prompt_kwargs: dict, llm_kwargs: Optional[dict] = None):
+        combined = {"prompt_kwargs": prompt_kwargs, "llm_kwargs": llm_kwargs}
+        json_str = json.dumps(combined, sort_keys=True)
+        hash_obj = hashlib.sha256(json_str.encode("utf-8"))
+        return hash_obj.hexdigest()
+
     def generate(self, *, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None) -> Any:
+        if self._cache:
+            key = self._get_cache_key(prompt_kwargs, llm_kwargs)
+            hit = self._cache.get(key)
+            if hit:
+                if hit.get("prompt_kwargs") == prompt_kwargs and hit.get("llm_kwargs") == llm_kwargs:
+                    return hit.get("result")
+                else:
+                    logger.warning(
+                        "Found cache content mismatch, key=%s prompt_kwargs=%s llm_kwargs=%s",
+                        key,
+                        prompt_kwargs,
+                        llm_kwargs,
+                    )
+
         if llm_kwargs is not None:
-            return self._generate_using_openai(prompt_kwargs, llm_kwargs)
+            result = self._generate_using_openai(prompt_kwargs, llm_kwargs)
         else:
-            return self._generate_using_guidance(prompt_kwargs)
+            result = self._generate_using_guidance(prompt_kwargs)
+
+        if self._cache:
+            key = self._get_cache_key(prompt_kwargs, llm_kwargs)
```

Comment:
we can use the key generated above!

Reply:
They're both in conditional ops so the IDE was complaining :B Let me cleanup though